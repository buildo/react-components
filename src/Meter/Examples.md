### Examples

You can customize the style (colors) of Meter based on the range in which the current value falls

```js
const ranges = [
  {
    startValue: 0,
    endValue: 30,
    fillingColor: '#ed1c24',
    labelColor: '#ed1c24',
    backgroundColor: '#feeced'
  },
  { startValue: 30, endValue: 70, fillingColor: '#fdc018', labelColor: '#fdc018' },
  { startValue: 70, endValue: 100, fillingColor: '#34aa44', labelColor: '#34aa44' }
];

const meterProps = {
  ranges,
  baseBackgroundColor: '#f8f8f8',
  style: { marginBottom: 20 }
};

<FlexView width="100%" column>
  <Meter {...meterProps} value={30} />
  <Meter {...meterProps} value={45} />
  <Meter {...meterProps} value={100} />
</FlexView>;
```

Ranges could also have holes between them. If the current value doesn't fall in any of the given ranges, it will use a base style you can configure as you like:

```js
const ranges = [{ startValue: 40, endValue: 60, fillingColor: '#34aa44', labelColor: '#34aa44' }];

const meterProps = {
  ranges,
  baseFillingColor: '#CCCCCC',
  baseBackgroundColor: '#F8F8F8',
  style: { marginBottom: 20 }
};

<FlexView column>
  <Meter {...meterProps} value={20} />
  <Meter {...meterProps} value={50} />
  <Meter {...meterProps} value={80} />
</FlexView>;
```

#### Custom labels

By default, Meter display an auto-calculated percentage next to the bar. You can customize your own label formatter in order to show whathever you like:

```js
function labelFormatter1(curr, min, max) {
  return +(Math.round(curr + 'e+2') + 'e-2');
}

function labelFormatter2(curr, min, max) {
  return `${curr}/${max}`;
}

function labelFormatter3(curr, min, max) {
  return curr > (max - min) / 2 ? 'good' : 'bad';
}

const ranges = [
  { startValue: 400, endValue: 665, fillingColor: '#34aa44', labelColor: '#34aa44' },
  { startValue: 665, endValue: 667, fillingColor: '#ed1c24', labelColor: '#ed1c24' },
  { startValue: 667, endValue: 700, fillingColor: '#fdc018', labelColor: '#fdc018' }
];

const meterProps = {
  min: 400,
  max: 700,
  ranges,
  baseBackgroundColor: '#f8f8f8',
  style: { marginBottom: 20 }
};

<FlexView column>
  <Meter {...meterProps} value={432.315} labelFormatter={labelFormatter1} />
  <Meter {...meterProps} value={666} labelFormatter={labelFormatter2} />
  <Meter {...meterProps} value={690} labelFormatter={labelFormatter3} />
</FlexView>;
```
