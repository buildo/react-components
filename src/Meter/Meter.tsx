import * as React from 'react';
import { props, t } from '../utils';
import { warn } from '../utils/log';
import cx from '../utils/classnames';
import FlexView from 'react-flexview';
import find = require('lodash/find');
import every = require('lodash/every');
import isEqual = require('lodash/isEqual');
import sortBy = require('lodash/sortBy');

const Range = t.refinement(t.struct({
  startValue: t.Number,
  endValue: t.Number,
  fillingColor: t.maybe(t.String),
  labelColor: t.maybe(t.String),
  backgroundColor: t.maybe(t.String)
}), (r: any) => r.startValue < r.endValue, 'Range'); // FIXME: any

const Ranges = t.refinement(t.list(Range), (rangeList) => {

  const isOverlapped = (range1: MeterProps.Range, range2: MeterProps.Range) => (
    Math.max(range1.startValue, range2.startValue) < Math.min(range1.endValue, range2.endValue)
  );

  const noOverlappingRanges = (ranges: MeterProps.Range[]) => {
    return every(ranges, (range1, i) => {
      return every(ranges.slice(0, i).concat(ranges.slice(i + 1)), range2 => {
        return !isOverlapped(range1, range2);
      });
    });
  };

  return noOverlappingRanges(rangeList);
}, 'Ranges');

export const Props = t.refinement(t.struct({
  value: t.Number,
  min: t.maybe(t.Number),
  max: t.maybe(t.Number),
  labelFormatter: t.maybe(t.Function),
  ranges: t.maybe(Ranges),
  baseLabelColor: t.maybe(t.String),
  baseFillingColor: t.maybe(t.String),
  baseBackgroundColor: t.maybe(t.String),
  id: t.maybe(t.String),
  className: t.maybe(t.String),
  style: t.maybe(t.Object)
}), ({ min, max }) => min < max, 'Props');

const isFullyFilled = (ranges: MeterProps.Range[] | void, min: number, max: number) => {
  if (!ranges) {
    return false;
  }
  const comparableRanges = ranges.concat({ startValue: max, endValue: min });
  const sortedStartValueList = sortBy(comparableRanges, 'startValue').map(range => range.startValue);
  const sortedEndValueList = sortBy(comparableRanges, 'endValue').map(range => range.endValue);
  return isEqual(sortedStartValueList, sortedEndValueList);
};

const computePercentage = (value: number, min: number, max: number) => (
  Math.abs((value - min) * 100 / (max - min))
);

const labelFormatter = (value: number, min: number, max: number) => (
  `${computePercentage(value, min, max)}%`
);

export namespace MeterProps {
  export type Range = {
    startValue: number,
    endValue: number,
    fillingColor?: string,
    labelColor?: string,
    backgroundColor?: string
  }
};

export type MeterDefaultProps = {
  /** minimum value */
  min: number,
  /** maximum value */
  max: number,
  /** function in which you can define a custom label format */
  labelFormatter: (value: number, min: number, max: number) => string,
};

export type MeterRequiredProps = {
  /** current value */
  value: number,
  /** array of Object in which you can define startValue, endValue, labelColor, fillingColor */
  ranges?: MeterProps.Range[],
  /** fallback labelColor */
  baseLabelColor?: string,
  /** fallback fillingColor */
  baseFillingColor?: string,
  /** fallback backgroundColor */
  baseBackgroundColor?: string,
  id?: string,
  className?: string,
  style?: React.CSSProperties
};

export type MeterProps = Partial<MeterDefaultProps> & MeterRequiredProps;

const defaultProps = {
  min: 0,
  max: 100,
  labelFormatter
};

/**
 * A Meter displays a measurement (usually a percentage) on a known scale.
 */
@props(Props)
export default class Meter extends React.Component<MeterProps> {

  componentDidMount() {
    this.logWarnings();
  }

  logWarnings = () => {
    const {
      ranges,
      max,
      min,
      baseFillingColor,
      baseBackgroundColor
    } = this.getProps();
    warn(() => {
      if (isFullyFilled(ranges, min, max) && (baseFillingColor || baseBackgroundColor)) {
        return 'baseFillingColor or baseBackgroundColor is not needed, ranges are fully filled';
      }
      return undefined;
    });
    warn(() => {
      if (!(isFullyFilled(ranges, min, max) || (baseFillingColor && baseBackgroundColor))) {
        return 'You should pass both baseFillingColor and baseBackgroundColor, ranges are not fully filled';
      }
      return undefined;
    });
  };

  computeStyles = () => {
    const {
      value,
      min,
      max,
      ranges,
      baseFillingColor,
      baseLabelColor,
      baseBackgroundColor
    } = this.getProps();

    const range = find(ranges, ({ startValue, endValue }) => (
      value >= startValue && value <= endValue
    ));
    return {
      basisSize: `${computePercentage(value, min, max)}%`,
      fillingStyle: {
        backgroundColor: range ? range.fillingColor : baseFillingColor
      },
      labelStyle: {
        color: range ? range.labelColor : baseLabelColor
      },
      barStyle: {
        backgroundColor: range ? range.backgroundColor : baseBackgroundColor
      }
    };
  }

  getProps() {
    return { ...defaultProps, ...this.props };
  }

  render() {
    const {
      id,
      style,
      className: _className,
      labelFormatter,
      value,
      min,
      max
    } = this.getProps();

    const { basisSize, fillingStyle, labelStyle, barStyle } = this.computeStyles();
    const className = cx('meter', _className);
    const formattedLabel = labelFormatter(value, min, max);

    return (
      <FlexView {...{ id, className, style }} grow>
        <FlexView
          className='bar'
          grow
          style={barStyle}
        >
          <FlexView
            className='filling'
            basis={basisSize}
            style={fillingStyle}
          />
        </FlexView>
        <FlexView
          className='label'
          shrink={false}
          style={labelStyle}
        >
          {formattedLabel}
        </FlexView>
      </FlexView>
    );
  }

}
