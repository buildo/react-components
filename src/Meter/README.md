A Meter is a simple UI component used to display a measurement (usually a percentage) on a known scale.

![cover](https://user-images.githubusercontent.com/10867086/36546429-4d6179f2-17e3-11e8-8cc7-04d362304f7c.jpg)

Typically, there are two use cases for this component:
- display a value, such as a battery charge indicator
- display the progress of an action
# Usage
## Display a value

In this use case, the value can go up and down based on the semantics of the real world object.

![displayavalue](https://user-images.githubusercontent.com/10867086/36546866-5c9e3efe-17e4-11e8-9b7a-4ede95f0faa3.jpg)

For example, the battery charge goes up while recharging, and down while using the phone. Regarding ranges, there might be a "low battery" range in red between 0% and 15%, and a "normal" battery range between 15% and 100%.

If the meter is displaying the space occupied, the semantics are inverted: it will become red as it approaches 100% or the maximum space value in Gigabytes.

## Display the progress

In this use case, the value starts from 0 and can only go up, typically to 100%.

![progressbar](https://user-images.githubusercontent.com/10867086/36546608-c77eb2fe-17e3-11e8-9f72-5bac98150877.jpg)

Follow these high-level rules:
- do not show a progress indicator for actions that typically require less than a second
  - *example: loading a simple modal view*
- show a simple indicator (no label) for actions that typically require less than 10 seconds
  - *example: preparing a PDF document for download*
- show a full indicator with the percentage for longer actions and, if possible, a time estimate
  - *example: upgrading an on-premises application via a web admin console*

Only use one color for the progress indicator.

To reassure users that the app is doing something, the progress bar should never stop. To do this, try to start slow and accelerate when you have real progress to show.
