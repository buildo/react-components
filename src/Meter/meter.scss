@import '../utils/theme/variables.scss';

$height: 18px !default;
$font-size: 14px !default;
$font-weight: $brc-regular !default;
$border-radius: 10px !default;
$label-margin: 5px !default;

.meter {
  .bar {
    overflow: hidden;
    border-radius: $border-radius;
    height: $height;

    .filling {
      border-bottom-left-radius: $border-radius;
      border-top-left-radius: $border-radius;
    }
  }

  .label {
    font-size: $font-size;
    font-weight: $font-weight;
    padding-left: $label-margin;
  }
}
