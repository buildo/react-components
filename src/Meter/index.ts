import Meter from './Meter';
import { MeterProps } from './Meter';
export default Meter;
export { MeterProps };
