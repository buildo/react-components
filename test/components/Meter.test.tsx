import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import Meter, { MeterProps } from '../../src/Meter';

const style: React.CSSProperties = {
  margin: 10,
  position: 'relative'
};

const ranges: MeterProps.Range[] = [
  { startValue: 0, endValue: 50, fillingColor: 'green', labelColor: 'red', backgroundColor: 'grey' },
  { startValue: 50, endValue: 80, fillingColor: 'yellow' },
  { startValue: 80, endValue: 100 }
];

const exampleProps = {
  id: '12345',
  className: 'fancy-class-name',
  style,
  value: 25,
  min: 0,
  max: 50,
  ranges,
  baseLabelColor: '#000000',
  baseFillingColor: '#ccc',
  labelFormatter: () => 'test'
};

const meter = shallow(<Meter {...exampleProps} />);

describe('Meter', () => {

  it('renders correctly', () => {
    const component = renderer.create(
      <Meter {...exampleProps} />
    );
    expect(component).toMatchSnapshot();
  });

  it('computes className', () => {
    expect(meter.hasClass('meter')).toBe(true);
    expect(meter.hasClass('fancy-class-name')).toBe(true);
  });

  it('uses labelFormatter', () => {
    expect(meter.find('.label').dive().text()).toBe('test');
  });

  it('computes barStyle', () => {
    expect(meter.find('.bar').prop('style').backgroundColor).toBe(ranges[0].backgroundColor);
  });

  it('computes fillingStyle', () => {
    expect(meter.find('.filling').prop('style').backgroundColor).toBe(ranges[0].fillingColor);
  });

  it('computes labelStyle', () => {
    expect(meter.find('.label').prop('style').color).toBe(ranges[0].labelColor);
  });

  it('computes basisSize', () => {
    expect(meter.find('.filling').prop('basis')).toBe('50%');
  });

  it('computes basisSize with custom min and max', () => {
    const meter = shallow(
      <Meter value={150} min={100} max={200} />
    );
    expect(meter.find('.filling').prop('basis')).toBe('50%');
  });

  it('computes basisSize for custom negative min', () => {
    const meter = shallow(
      <Meter value={0} min={-100} max={100} />
    );
    expect(meter.find('.filling').prop('basis')).toBe('50%');
  });

  it('computes background color for filling', () => {
    const ranges = [
      { startValue: 50, endValue: 80, fillingColor: 'yellow' }
    ];
    const meter = shallow(
      <Meter value={60} ranges={ranges} />
    );
    expect(meter.find('.filling').prop('style').backgroundColor).toBe(ranges[0].fillingColor);
  });

  it('defaults to the base color as background color if there\'s no matching range', () => {
    const ranges = [
      { startValue: 50, endValue: 80, fillingColor: 'yellow' }
    ];
    const meter = shallow(
      <Meter value={20} ranges={ranges} baseFillingColor='#ccc' />
    );
    expect(meter.find('.filling').prop('style').backgroundColor).toBe('#ccc');
  });

  it('doesn\'t define a background color if there\'s no matching range and no default is given', () => {
    const ranges = [
      { startValue: 50, endValue: 80, fillingColor: 'yellow' }
    ];
    const meter = shallow(
      <Meter value={20} ranges={ranges} />
    );
    expect(meter.find('.filling').prop('style').backgroundColor).toBeUndefined();
  });

  it('defaults to base color as bar background color should if there\'s no matching range', () => {
    const ranges = [
      { startValue: 50, endValue: 80, fillingColor: 'yellow' }
    ];
    const meter = shallow(
      <Meter value={20} ranges={ranges} baseBackgroundColor='#ccc' />
    );
    expect(meter.find('.bar').prop('style').backgroundColor).toBe('#ccc');
  });

  it('doesn\'t define a bar background color if there\'s no matching range and no default is given', () => {
    const ranges = [
      { startValue: 50, endValue: 80, fillingColor: 'yellow' }
    ];
    const meter = shallow(
      <Meter value={20} ranges={ranges} />
    );
    expect(meter.find('.bar').prop('style').backgroundColor).toBeUndefined();
  });

});
