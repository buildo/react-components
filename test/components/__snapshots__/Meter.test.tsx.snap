// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Meter renders correctly 1`] = `
<div
  className="react-flex-view meter fancy-class-name"
  id="12345"
  style={
    Object {
      "MozBoxFlex": "1 1 auto",
      "WebkitBoxFlex": "1 1 auto",
      "WebkitFlex": "1 1 auto",
      "flex": "1 1 auto",
      "margin": 10,
      "msFlex": "1 1 auto",
      "position": "relative",
    }
  }
>
  <div
    className="react-flex-view bar"
    style={
      Object {
        "MozBoxFlex": "1 1 auto",
        "WebkitBoxFlex": "1 1 auto",
        "WebkitFlex": "1 1 auto",
        "backgroundColor": "grey",
        "flex": "1 1 auto",
        "msFlex": "1 1 auto",
      }
    }
  >
    <div
      className="react-flex-view filling"
      style={
        Object {
          "MozBoxFlex": "0 0 50%",
          "WebkitBoxFlex": "0 0 50%",
          "WebkitFlex": "0 0 50%",
          "backgroundColor": "green",
          "flex": "0 0 50%",
          "msFlex": "0 0 50%",
        }
      }
    />
  </div>
  <div
    className="react-flex-view label"
    style={
      Object {
        "MozBoxFlex": "0 0 auto",
        "WebkitBoxFlex": "0 0 auto",
        "WebkitFlex": "0 0 auto",
        "color": "red",
        "flex": "0 0 auto",
        "msFlex": "0 0 auto",
      }
    }
  >
    test
  </div>
</div>
`;
